eth/tracers/native/call_flat.go

```diff
@@ -0,0 +1,346 @@
+	t.tracer.CaptureFault(pc, op, gas, cost, scope, depth, err)
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (t *flatCallTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureEnter(typ, from, to, input, gas, value)
+}
+
+// CaptureExit is called when EVM exits a scope, even if the scope didn't
+// execute any code.
+func (t *flatCallTracer) CaptureExit(output []byte, gasUsed uint64, err error) {
+	t.tracer.CaptureExit(output, gasUsed, err)
+}
+
+func (t *flatCallTracer) CaptureTxStart(gasLimit uint64) {
+	t.tracer.CaptureTxStart(gasLimit)
+}
+
+func (t *flatCallTracer) CaptureTxEnd(restGas uint64) {
+	t.tracer.CaptureTxEnd(restGas)
+}
+
+// GetResult returns an empty json object.
+func (t *flatCallTracer) GetResult() (json.RawMessage, error) {
+	traceResultJson, err := t.tracer.GetResult()
+	if err != nil {
+		return nil, err
+	}
+
+	traceResult := new(callFrame)
+	err = json.Unmarshal(traceResultJson, &traceResult)
+	if err != nil {
+		return nil, err
+	}
+
+	flat, err := t.processOutput(traceResult, []int{})
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := json.Marshal(flat)
+	if err != nil {
+		return nil, err
+	}
+	return json.RawMessage(res), t.reason
+}
+
+// Stop terminates execution of the tracer at the first opportune moment.
+func (t *flatCallTracer) Stop(err error) {
+	t.tracer.Stop(err)
+}
+
+func (t *flatCallTracer) processOutput(input *callFrame, traceAddress []int) (output []flatCallFrame, err error) {
+	gasHex := hexutil.Uint64(input.Gas)
+	gasUsedHex := hexutil.Uint64(input.GasUsed)
+	valueHex := hexutil.Big{}
+	if input.Value != nil {
+		valueHex = hexutil.Big(*input.Value)
+	}
+
+	frame := flatCallFrame{
+		Type: strings.ToLower(input.Type.String()),
+		Action: FlatCallTraceAction{
+			From:  &input.From,
+			To:    &input.To,
+			Gas:   &gasHex,
+			Value: &valueHex,
+		},
+		Result: &FlatCallTraceResult{
+			GasUsed: &gasUsedHex,
+		},
+		Error:        input.Error,
+		TraceAddress: traceAddress,
+	}
+
+	// typ := vm.StringToOp(strings.ToUpper(call.Type))
+	if input.Type == vm.CREATE || input.Type == vm.CREATE2 {
+		t.formatCreateResult(&frame, input)
+	} else if input.Type == vm.SELFDESTRUCT {
+		t.formatSuicideResult(&frame, input)
+	} else {
+		t.formatCallResult(&frame, input)
+	}
+
+	t.fillCallFrameFromContext(&frame)
+
+	if t.config.ConvertedParityErrors {
+		t.convertErrorToParity(&frame)
+	}
+
+	frame.Subtraces = len(input.Calls)
+
+	output = append(output, frame)
+
+	if len(input.Calls) > 0 {
+		for i, childCall := range input.Calls {
+			traceAddress = append(traceAddress, i)
+			flat, err := t.processOutput(&childCall, traceAddress)
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, flat...)
+		}
+	}
+
+	return output, nil
+}
+
+func (t *flatCallTracer) fillCallFrameFromContext(callFrame *flatCallFrame) {
+	if t.ctx != nil {
```

Comment:
Hm I notice the context doesn't have the block number which is usually part of the parity trace. I'll have to look into why

Reply:
It has to be added, as said the PR is not finished.
I would like to know your thoughts on using a proxy tracers to `call_tracer` or a new tracer for flat_tracer.

The one thing I liked with the proxy one, is that we will have a single point in codebase to maintain the tracer logic. Although this approach, has some nitpicks in order to work.

Example nitpick which is not pushed here follows. Here and because we don't have access to `callTracer.callstack` we have to keep track of some stuff in this tracer ourselves. Making the `callTracer.**C**allstack` public will help, but then it will make more trivial to maintain the **callTracer** if we want to apply any changes.

```go
// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
func (t *flatCallTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
	// Update list of precompiles based on current block
	rules := env.ChainConfig().Rules(env.Context.BlockNumber, env.Context.Random != nil)
	t.activePrecompiles = vm.ActivePrecompiles(rules)

	// this is a workaround in order to exclude precompile calls in CaptureExit where we don't have the proxied tracer context
	t.currentCall = flatCallFrame{
		Type: strings.ToLower(vm.CALL.String()),
		Action: FlatCallTraceAction{
			From: &from,
			To:   &to,
			// Gas:   gas,
			// Value: value,
		},
	}

	if create {
		t.currentCall.Type = strings.ToLower(vm.CREATE.String())
	}

	t.tracer.CaptureStart(env, from, to, create, input, gas, value)
}
```

Reply:
> is that we will have a single point in codebase to maintain the tracer logic

I agree with that. As long as there aren't many work-arounds needed because of differences between parity-vs-geth styles like the one mentioned. That depends on whether we want a flatTracer or we want the parityTracer. Because as you mentioned parity tracer has some non-cosmetic differences like precompiles being excluded. Do you know what are the other differences?

You can cast the `t.tracer.(*callTracer)` and then you'll have access to `callstack` since they're in the same package. That should make things easier.

Comparing this version and the one in `core-geth` I must admit I like this one more. Even if we don't go with the proxy approach I think we should track the calls in the same shape that they will be returned (i.e. flat) so we don't need post-processing apart from toHex and toJSON conversions. I.e. avoid the `Finalize` method in your code. I realize a challenge is that we need access to the call in `CaptureExit`. But we can find the index of the call in the flat list by storing an additional stack that only tracks indices.

Reply:
> Because as you mentioned parity tracer has some non-cosmetic differences like precompiles being excluded.

Do you find a bad idea to add an option in **CallTracer** that allows user to decide? `excludePrecompiles: boolean`

> You can cast the t.tracer.(*callTracer) and then you'll have access to callstack since they're in the same package. That should make things easier.

Nice one. Will check it and think if it adds more complexity than needed for maintenance later on.